sdk/python/kfp/dsl/component_spec.py

```diff
@@ -107,38 +110,118 @@ def build_task_inputs_spec(
     task_spec: pipeline_spec_pb2.PipelineTaskSpec,
     pipeline_params: List[_pipeline_param.PipelineParam],
     tasks_in_current_dag: List[str],
+    is_parent_component_root: bool,
 ) -> None:
   """Builds task inputs spec from pipeline params.
 
   Args:
     task_spec: The task spec to fill in its inputs spec.
     pipeline_params: The list of pipeline params.
     tasks_in_current_dag: The list of tasks names for tasks in the same dag.
+    is_parent_component_root: Whether the task is in the root component.
   """
   for param in pipeline_params or []:
-    input_name = param.full_name
+
+    input_name = _pipeline_param.additional_input_name_for_pipelineparam(param)
     if type_utils.is_parameter_type(param.param_type):
-      if param.op_name in tasks_in_current_dag:
+      if param.op_name and dsl_utils.sanitize_task_name(
+          param.op_name) in tasks_in_current_dag:
         task_spec.inputs.parameters[
             input_name].task_output_parameter.producer_task = (
                 dsl_utils.sanitize_task_name(param.op_name))
         task_spec.inputs.parameters[
             input_name].task_output_parameter.output_parameter_key = (
                 param.name)
       else:
-        task_spec.inputs.parameters[
-            input_name].component_input_parameter = input_name
+        task_spec.inputs.parameters[input_name].component_input_parameter = (
+            param.full_name if is_parent_component_root else input_name)
     else:
-      if param.op_name in tasks_in_current_dag:
+      if param.op_name and dsl_utils.sanitize_task_name(
+          param.op_name) in tasks_in_current_dag:
         task_spec.inputs.artifacts[
             input_name].task_output_artifact.producer_task = (
                 dsl_utils.sanitize_task_name(param.op_name))
         task_spec.inputs.artifacts[
             input_name].task_output_artifact.output_artifact_key = (
                 param.name)
       else:
-        task_spec.inputs.artifacts[
-            input_name].component_input_artifact = input_name
+        task_spec.inputs.artifacts[input_name].component_input_artifact = (
+            param.full_name if is_parent_component_root else input_name)
+
+
+def update_task_inputs_spec(
```

Comment:
Is this a public user-visible API?

Reply:
Kind of. No need to change it in this PR.
Perhaps we can later rename the whole file with `_` prefix (like the other files in dsl/).